packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -557,162 +558,163 @@ impl<'m> RunnerImpl<'m> {
+                // a start offset and an end offset. The start offset is equal to the end offset of
+                // the previous value, thus we need `num_values + 1` offset values.
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow as offsets, the type is `i32`,
+                //   and `target_len + 1` is carefully chosen.
+                let offsets = unsafe {
+                    slice::from_raw_parts(data.buffer_ptrs[0] as *const i32, target_len + 1)
+                };
+                builder = builder.add_buffer(Buffer::from_slice_ref(&offsets));
+
+                // SAFETY: `data.buffer_ptrs[1]` is provided by arrow as values, the type is
+                //   `u8`, and the length is provided by the offsets
+                let values = unsafe {
+                    slice::from_raw_parts(data.buffer_ptrs[1], offsets[target_len] as usize)
+                };
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            }
+            DataType::List(inner_field) => {
+                // List is stored in one buffer and child data containing the indexed values:
+                //   buffer: The offset buffer (i32)
+                //   child_data: The value data
+
+                // See `DataType::Utf8` above for reasoning on `target_len + 1`
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `i32`, and
+                //   `target_len + 1` is carefully chosen.
+                let offsets = unsafe {
+                    slice::from_raw_parts(data.buffer_ptrs[0] as *const i32, target_len + 1)
+                };
+                builder = builder.add_buffer(Buffer::from_slice_ref(&offsets));
+
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                let child = child_data.get(0)?;
+                builder = builder.add_child_data(self.array_data_from_js(
                     mv8,
                     &child,
-                    field.data_type(),
+                    inner_field,
                     None,
-                )?])
+                )?);
             }
-            DataType::FixedSizeList(field, multiplier) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+            DataType::FixedSizeList(inner_field, size) => {
+                // FixedSizeListList is only stored by child data, as offsets are not required
+                // because the size is known.
+                let child_data: mv8::Array<'_> = obj.get("child_data")?;
+                let child = child_data.get(0)?;
+                builder = builder.add_child_data(self.array_data_from_js(
                     mv8,
                     &child,
-                    field.data_type(),
-                    Some(data.len * multiplier as usize),
-                )?])
+                    inner_field,
+                    Some(target_len * *size as usize),
```

Comment:
```suggestion
                    Some(*size as usize * target_len),
```
is a little more readable